### PR TITLE
Add ls command to cxflow CLI

### LINKED ISSUE #36: Add ls command to cxflow for simple training listing

### PR DESCRIPTION
**Pull request by blazekadam**
## changes
- add ls command to cxflow CLI
- move `CXF_*` variables to a separate file
- include date and milliseconds in log entries
- log and config file names are now `CXF` constants
- add epochs done logging to cxflow `main_loop`

## `cxflow ls` docstring
```
    List training dirs contained in the given dir with options and outputs similar to the regular `ls` command.

    The function is accessible through cxflow CLI `cxflow ls`.

    :param dir_: dir to be listed
    :param recursive: walk recursively in sub-directories, stop at train dirs (--recursive option)
    :param all_: include train dirs with no epochs done (--all option)
    :param long: list more more details including net name, net and dataset class/modules,
                 age, duration and epochs done (--long option)
    :param verbose: print more verbose output with list of additional artifacts and training config,
                    applicable only when a single train dir si listed (--verbose option)
```

## `cxflow ls` examples
The following command examples are from my local isletnet project dir with
- log dir containing 100+ training dirs with the old log formats (i.e., the command can not derive epochs done, and timings; hence, they are included only when --all option is specified) and 1 training with the new log format
- one training dir with the new log format in the project root

Command:
```
cxflow ls
```
Output:
```
IsletNet_2017-07-26-00-34-11_2hous1d4
```

____

Command:
```
cxflow ls -la
```
Output:
```
total 134
IsletNetCNTK_2017-06-28-00-55-38__c6_r4dg  IsletNetCNTK  nets.IsletNetCNTK  datasets.IsletNetDataset  N/A        N/A        N/A
IsletNetCNTK_2017-06-28-00-56-55_3zbpx43s  IsletNetCNTK  nets.IsletNetCNTK  datasets.IsletNetDataset  N/A        N/A        N/A
IsletNetCNTK_2017-06-28-01-02-02_583kl5ug  IsletNetCNTK  nets.IsletNetCNTK  datasets.IsletNetDataset  N/A        N/A        N/A
IsletNetCNTK_2017-06-28-01-03-17_w44437u0  IsletNetCNTK  nets.IsletNetCNTK  datasets.IsletNetDataset  N/A        N/A        N/A
...
IsletNet_2017-07-25-22-59-21_1kqkqvu2      IsletNet      nets.IsletNet      datasets.IsletNetDataset  N/A        N/A        N/A
IsletNet_2017-07-25-23-01-36_1dvx0lv7      IsletNet      nets.IsletNet      datasets.IsletNetDataset  N/A        N/A        N/A
IsletNet_2017-07-25-23-48-58_5lq8k5e5      IsletNet      nets.IsletNet      datasets.IsletNetDataset  1 day ago  2 minutes  N/A
IsletNet_2017-07-26-00-34-11_2hous1d4      IsletNet      nets.IsletNet      datasets.IsletNetDataset  1 day ago  3 minutes  2

+---------+
| summary |
+---------+

+--------------+---------+
| net.name     |   count |
+==============+=========+
| IsletNetCNTK |     122 |
+--------------+---------+
| IsletNet     |      12 |
+--------------+---------+

+--------------------+--------------------------+---------+
| net.module.class   | dataset.module.class     |   count |
+====================+==========================+=========+
| nets.IsletNetCNTK  | datasets.IsletNetDataset |     122 |
+--------------------+--------------------------+---------+
| nets.IsletNet      | datasets.IsletNetDataset |      12 |
+--------------------+--------------------------+---------+

```

____

Command:
```
cxflow ls -lr .
```
Output:
```
.:
total 1
IsletNet_2017-07-26-00-43-44_ofwed8x_  IsletNet  nets.IsletNet  datasets.IsletNetDataset  1 day ago  12 minutes  8

./log:
total 1
IsletNet_2017-07-26-00-34-11_2hous1d4  IsletNet  nets.IsletNet  datasets.IsletNetDataset  1 day ago  3 minutes  2

+---------+
| summary |
+---------+

+------------+---------+
| net.name   |   count |
+============+=========+
| IsletNet   |       2 |
+------------+---------+

+--------------------+--------------------------+---------+
| net.module.class   | dataset.module.class     |   count |
+====================+==========================+=========+
| nets.IsletNet      | datasets.IsletNetDataset |       2 |
+--------------------+--------------------------+---------+

```

____

Command:
```
cxflow ls -v log/IsletNet_2017-07-26-00-34-11_2hous1d4 
```
Output:
```
total 1
IsletNet_2017-07-26-00-34-11_2hous1d4  IsletNet  nets.IsletNet  datasets.IsletNetDataset  1 day ago  3 minutes  2

+--------+
| config |
+--------+
dataset: {batch_size: 1, data_root: islets_data, dataset_class: IsletNetDataset, dataset_module: datasets,
  flip: true, rotate: true, split: ostro-split.csv}
hooks:
- class: StatsHook
  variables:
    f1: [mean]
    loss: [mean]
    precision: [mean]
    recall: [mean]
- {class: CSVHook}
- {class: LoggingHook}
- {class: ProfileHook}
- {class: SigintHook}
- {class: BestSaverHook}
- {class: LRDecayHook, lr_multiplier: 0.997, module: cxflow_tf}
main_loop:
  extra_streams: [valid, test]
  skip_zeroth_epoch: true
net:
  channels: [8, 8, 12, 12, 24, 24, 28, 28, 32, 32]
  convs: [3, 5, 3, 5, 3, 7, 3, 5, 3, 5]
  device: /gpu:0
  io:
    in: [images, labels]
    out: [loss, predictions, f1, precision, recall]
  name: IsletNet
  net_class: IsletNet
  net_module: nets
  optimizer: {class: RMSPropOptimizer, learning_rate: 0.00015}
  pools: [1, 1, 1, 2, 1, 1, 1, 1, 1, 1]


+-----------+
| artifacts |
+-----------+
-  split.csv                            645.0  B
-  training.csv                         361.0  B
-  model_best.pb                        848.0  KB
-  model_best.graph                     251.9  KB
-  model_best.ckpt.data-00000-of-00001    2.4  MB
-  model_best.ckpt.index                  4.5  KB
-  checkpoint                            87.0  B
-  model_best.ckpt.meta                 327.1  KB

```

closes #36

### CODE REVIEW

File: cxflow/entry_point.py [hunk 1]
```diff
@@ -20,21 +20,15 @@
 from datetime import datetime
 from typing import Iterable, Optional
 
+from .constants import *
```
Comment (FloopCZ) [inline cxflow/entry_point.py]: Doesn't this reduce your pylint score? :-D
I would keep it anyway, though.

File: cxflow/entry_point.py [hunk 2]
```diff
@@ -337,6 +331,20 @@ def entry_point() -> None:
     subparsers = main_parser.add_subparsers(help='cxflow modes')
 
     # create train subparser
+    ls_parser = subparsers.add_parser('ls')
+    ls_parser.set_defaults(subcommand='ls')
+    ls_parser.add_argument('dir', nargs='?', default=CXF_DEFAULT_LOG_DIR,
+                           help='path to the log directory to be listed')
+    ls_parser.add_argument('-l', '--long', default=False, action='store_true',
```
Comment (FloopCZ) [inline cxflow/entry_point.py]: The default is not needed, the `action='store_true'` takes care of it (see the [docs](https://docs.python.org/3/library/argparse.html)).
The same stands for all the other defaults below.

**Comment by blazekadam** [comment]
Yet another (quite useful example of usage)

Command:
```
cxflow ls -la ~
```
Output:
```
total 10
slim-autoenc-90f1_2                            OstroFCN                       nets.ostro_fcn.AutoencoderNet  datasets.ostro_dataset.OstroDataset  N/A  N/A  N/A
slim-autoenc-90f1                              OstroFCN                       nets.ostro_fcn.AutoencoderNet  datasets.ostro_dataset.OstroDataset  N/A  N/A  N/A
dilated_weighing_2017-06-12-21-25-11_zdirb99a  dilated_weighing               nets.ostro_fcn.AutoencoderNet  datasets.ostro_dataset.OstroDataset  N/A  N/A  N/A
slim-autoenc-90f1_3                            dilated_weighing               nets.ostro_fcn.AutoencoderNet  datasets.ostro_dataset.OstroDataset  N/A  N/A  N/A
slim-autoenc-90f1_4                            weighed                        nets.ostro_fcn.AutoencoderNet  datasets.ostro_dataset.OstroDataset  N/A  N/A  N/A
fold4                                          split4                         nets.ostro_fcn.AutoencoderNet  datasets.ostro_dataset.OstroDataset  N/A  N/A  N/A
fold2                                          lr_decay                       nets.ostro_fcn.AutoencoderNet  datasets.ostro_dataset.OstroDataset  N/A  N/A  N/A
fold1                                          wider_deeper_more_regularized  nets.ostro_fcn.AutoencoderNet  datasets.ostro_dataset.OstroDataset  N/A  N/A  N/A
fold3                                          split3                         nets.ostro_fcn.AutoencoderNet  datasets.ostro_dataset.OstroDataset  N/A  N/A  N/A
fold5                                          split5                         nets.ostro_fcn.AutoencoderNet  datasets.ostro_dataset.OstroDataset  N/A  N/A  N/A

+---------+
| summary |
+---------+

+-------------------------------+---------+
| net.name                      |   count |
+===============================+=========+
| OstroFCN                      |       2 |
+-------------------------------+---------+
| dilated_weighing              |       2 |
+-------------------------------+---------+
| weighed                       |       1 |
+-------------------------------+---------+
| split4                        |       1 |
+-------------------------------+---------+
| lr_decay                      |       1 |
+-------------------------------+---------+
| wider_deeper_more_regularized |       1 |
+-------------------------------+---------+
| split3                        |       1 |
+-------------------------------+---------+
| split5                        |       1 |
+-------------------------------+---------+

+-------------------------------+-------------------------------------+---------+
| net.module.class              | dataset.module.class                |   count |
+===============================+=====================================+=========+
| nets.ostro_fcn.AutoencoderNet | datasets.ostro_dataset.OstroDataset |      10 |
+-------------------------------+-------------------------------------+---------+
```

File: cxflow/constants.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+CXF_DEFAULT_LOG_DIR = './log'
+
+CXF_LOG_FORMAT = '%(asctime)s.%(msecs)06d: %(levelname)-8s@%(module)-15s: %(message)s'
+CXF_LOG_DATE_FORMAT = '%Y-%m-%d %H:%M:%S'
+CXF_FULL_DATE_FORMAT = '%Y-%m-%d %H:%M:%S.%f'
+
+CXF_LOG_FORMATTER = logging.Formatter(CXF_LOG_FORMAT, datefmt=CXF_LOG_DATE_FORMAT)
+# Module with standard cxflow hooks (as would be used in import).
+CXF_HOOKS_MODULE = 'cxflow.hooks'
+
```
Comment (FloopCZ) [inline cxflow/constants.py]: There is a trailing line at the end of the file (I know that previously, there was no newline, but now, there are two newlines).

Comment (blazekadam) [inline cxflow/constants.py]: Thanks, resolved

File: cxflow/entry_point.py (hunk 2 above)
Comment (blazekadam) [inline cxflow/entry_point.py]: Thanks, resolved

File: cxflow/utils/ls.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+        # try to parse datetimes from the end backwards until success to find the last entry datetime
+        lane_index = -1
+        while True:
+            try:
+                end_datetime = datetime.strptime(lanes[lane_index].split(': ', 1)[0], CXF_FULL_DATE_FORMAT)
+                break
+            except ValueError:
+                pass
+            lane_index -= 1
```
Comment (FloopCZ) [inline cxflow/utils/ls.py]: If there is no end datetime, this will go to negative numbers and fail with an exception on line 94.

File: cxflow/utils/ls.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+                duration = format_timedelta(end_datetime - start_datetime)
+            else:
+                duration = 'N/A'
+        else:
+            age = 'N/A'
+            duration = 'N/A'
+
+        epochs_done = get_epochs_done(log)
+        if not epochs_done:
+            epochs_done = 'N/A'  # pylint: disable=redefined-variable-type
```
Comment (FloopCZ) [inline cxflow/utils/ls.py]: instead of the pylint info here, you may consider to do `str(get_epochs_done(log))` (your choice)

File: cxflow/utils/ls.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+
+
+def _ls_print_listing(dir_: str, recursive: bool, all_: bool, long: bool) -> List[Tuple[str, dict, str]]:
+    """
+    Print names of the train dirs contained in the given dir.
+
+    :param dir_: dir to be listed
+    :param recursive: walk recursively in sub-directories, stop at train dirs (--recursive option)
+    :param all_: include train dirs with no epochs done (--all option)
+    :param long: list more more details including net name, net and dataset class/modules,
```
Comment (FloopCZ) [inline cxflow/utils/ls.py]: more more

File: cxflow/utils/ls.py [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+
+    The function is accessible through cxflow CLI `cxflow ls`.
+
+    :param dir_: dir to be listed
+    :param recursive: walk recursively in sub-directories, stop at train dirs (--recursive option)
+    :param all_: include train dirs with no epochs done (--all option)
+    :param long: list more more details including net name, net and dataset class/modules,
+                 age, duration and epochs done (--long option)
+    :param verbose: print more verbose output with list of additional artifacts and training config,
+                    applicable only when a single train dir si listed (--verbose option)
```
Comment (FloopCZ) [inline cxflow/utils/ls.py]: is listed

File: cxflow/utils/ls.py [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+
+    if long and len(all_trainings) > 1:
+        if not recursive:
+            print()
+        _ls_print_summary(all_trainings)
+
+    if verbose and len(all_trainings) == 1:
+        if not recursive:
+            print()
+        _ls_print_verbose(all_trainings[0])
```
Comment (FloopCZ) [inline cxflow/utils/ls.py]: I would not oppose to allow verbose output for more than one folder if they are explicitly specified in the command line arguments. (for instance, if I do `cxflow -v log/my-super-net-test-*`)

File: cxflow/utils/ls.py [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+from collections import defaultdict
+from typing import Iterable, Optional, Tuple, List
+
+from tabulate import tabulate
+from babel.dates import format_timedelta
+
+from .config import load_config, config_to_str
+from ..constants import CXF_DEFAULT_LOG_DIR, CXF_CONFIG_FILENAME, CXF_LOG_FILENAME, CXF_FULL_DATE_FORMAT
+
+CXF_NA_STR = 'N/A'
```
Comment (blazekadam) [inline cxflow/utils/ls.py]: unused indeed, not sure about moving it to `..constants` as it is rather a local constant. nonetheless, let's be consistent, I ll move it there

File: cxflow/utils/ls.py (hunk 8 above)
Comment (blazekadam) [inline cxflow/utils/ls.py]: At the moment, only one dir to be listed can be specified.

**Comment by blazekadam** [comment]
Thanks @FloopCZ for your comments. I fixed the obvious issues.

The main point to be discussed is whether log parsing is a good design patter and if not, how to avoid it. 

**My short answer is: It is not, yet we may live it with here.**

My thoughts behind this answer are:
- we parse the log only for age, duration and no. of epochs, it may fail of course (e.g. for backwards incompatibility reasons) which would produce N/As which is acceptable
- at the moment, we allow only INFO and DEBUG level logging (with --verbose flag); I would not allow any other levels at least for the file log. The consequence is that the parsed dates, duration etc. are always sound
- I would like to avoid creating any proprietary meta-info artifact/s only to hold this info, yet another thing to document, explain and confuse the user; furthermore, it would be something in between the two management options we agreed on (simple always-available ls and NoSQL management instruments covering detailed training info, losses, accuracy etc.)

File: cxflow/utils/ls.py [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+    total_size = 0
+    for root_dir, _, files in os.walk(path_):
+        for file_ in files:
+            total_size += path.getsize(path.join(root_dir, file_))
+    return total_size
+
+
+def humanize_filesize(filesize: int) -> Tuple[str, str]:
+    """Return human readable pair of size and unit from the given filesize in bytes."""
+    for unit in ['', 'K', 'M', 'G', 'T', 'P', 'E', 'Z']:
```
Comment (petrbel) [inline cxflow/utils/ls.py]: `k` should be lowercase
